### PR TITLE
Move actions, reducers, and configureStore to here from twreporter-react and @twreporter/universal-header 

### PR DESCRIPTION
**Pull request by YuCJ**
- twrepoter-react 中不需要再自己組 rootReducer 和 store，應該直接用 @twreporter/redux 的 `createStore()`，container 也應該要直接 call `store.actions`

Based on #23 

See details on Kanbanflow:
https://kanbanflow.com/t/egm2zAqE

### CODE REVIEW

**Comment by nickhsine** [comment]
LGTM

File: src/constants/action-types.js [hunk 1]
```diff
@@ -33,4 +33,86 @@ export default {
...
+  FETCH_ARTICLE_SUCCESS: 'FETCH_ARTICLE_SUCCESS',
+  FETCH_ARTICLE_FAILURE: 'FETCH_ARTICLE_FAILURE',
+
+  FETCH_RELATED_ARTICLES_REQUEST: 'FETCH_RELATED_ARTICLES_REQUEST',
+  FETCH_RELATED_ARTICLES_SUCCESS: 'FETCH_RELATED_ARTICLES_SUCCESS',
+  FETCH_RELATED_ARTICLES_FAILURE: 'FETCH_RELATED_ARTICLES_FAILURE',
+
+  FETCH_ARTICLES_BY_GROUP_UUID_REQUEST: 'FETCH_ARTICLES_BY_GROUP_UUID_REQUEST',
+  FETCH_ARTICLES_BY_GROUP_UUID_SUCCESS: 'FETCH_ARTICLES_BY_GROUP_UUID_SUCCESS',
+  FETCH_ARTICLES_BY_GROUP_UUID_FAILURE: 'FETCH_ARTICLES_BY_GROUP_UUID_FAILURE',
```
Comment (nickhsine) [inline src/constants/action-types.js]: delete `line 37` - `line 59` ?

File: src/constants/action-types.js [hunk 2]
```diff
@@ -33,4 +33,86 @@ export default {
...
+  LIST_ALL_AUTHORS_SUCCESS: 'LIST_ALL_AUTHORS_SUCCESS',
+  LIST_ALL_AUTHORS_FAILURE: 'LIST_ALL_AUTHORS_FAILURE',
+
+  FETCH_AUTHOR_COLLECTION_REQUEST: 'FETCH_AUTHOR_COLLECTION_REQUEST',
+  FETCH_AUTHOR_COLLECTION_FAILURE: 'FETCH_AUTHOR_COLLECTION_FAILURE',
+  FETCH_AUTHOR_COLLECTION_SUCCESS: 'FETCH_AUTHOR_COLLECTION_SUCCESS',
+
+  FETCH_FEATURE_ARTICLES_REQUEST: 'FETCH_FEATURE_ARTICLES_REQUEST',
+  FETCH_FEATURE_ARTICLES_SUCCESS: 'FETCH_FEATURE_ARTICLES_SUCCESS',
+  FETCH_FEATURE_ARTICLES_FAILURE: 'FETCH_FEATURE_ARTICLES_FAILURE',
```
Comment (nickhsine) [inline src/constants/action-types.js]: delete `line 77` - `line 79`?

File: src/constants/action-types.js [hunk 3]
```diff
@@ -33,4 +33,86 @@ export default {
...
+
+  SET_ARTICLE_TOPIC_LIST: 'SET_ARTICLE_TOPIC_LIST',
+  SET_BOOKMARKS_OF_LONGFORM_ARTICLE: 'SET_BOOKMARKS_OF_LONGFORM_ARTICLE',
+  SET_HEADER_INFO: 'SET_HEADER_INFO',
+  SET_PAGE_TITLE: 'SET_PAGE_TITLE',
+  SET_PAGE_THEME: 'SET_PAGE_THEME',
+  SET_PAGE_TYPE: 'SET_PAGE_TYPE',
+  SET_PROGRESS_PRECENTAGE: 'SET_PROGRESS_PRECENTAGE',
+
+  SET_ARTICLE_TOOLS: 'SET_ARTICLE_TOOLS',
```
Comment (nickhsine) [inline src/constants/action-types.js]: delete `line 81` - `line 89`?

File: src/reducers/header.js [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+      return {
+        ...state,
+        bookmarks: action.bookmarks,
+      }
+    default:
+      return state
+  }
+}
+
+export default header
```
Comment (nickhsine) [inline src/reducers/header.js]: delete this file?

File: src/reducers/index.js [hunk 5]
```diff
@@ -1,14 +1,59 @@
...
-import { posts } from './posts'
+import { post, posts } from './posts'
+import { searchedAuthorsList, authorsList } from './authors'
+import { topic, topics } from './topics'
+import authReducer from './auth'
+import bookmarks from './bookmarks'
+import bookmarkWidget from './bookmark-widget'
 import entities from './entities'
-import fieldNames from '../constants/redux-state-field-names'
+import header from './header'
```
Comment (nickhsine) [inline src/reducers/index.js]: remove header reducer?

File: src/actions/header.js [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+ * @param {string} info.pagetTopic - topic of article
+ * @param {string} info.pageType - type of page
+ * @param {array} info.topicArr - related articles in the same topic
+ */
+export function setHeaderInfo(info) {
+  return {
+    type: types.SET_HEADER_INFO,
+    info,
+  }
+}
```
Comment (nickhsine) [inline src/actions/header.js]: remove this file?

File: src/reducers/__test__/header.test.js [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+      })
+      expect(
+        reducer(
+          mockPreviousState,
+          mockActions[types.SET_BOOKMARKS_OF_LONGFORM_ARTICLE]
+        )
+      ).to.deep.equal(expectedStateOnPrev)
+    })
+  })
+})
```
Comment (nickhsine) [inline src/reducers/__test__/header.test.js]: delete this file?

File: src/actions/__test__/header.test.js [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+          pageTheme: 'mockPageTheme',
+          pagetTopic: 'mockPagetTopic',
+          pageType: 'mockPageType',
+          topicArr: ['mockTopic1', 'mockTopic2'],
+        },
+      }
+      expect(actions.setHeaderInfo(info)).to.deep.equal(expectedState)
+    })
+  })
+})
```
Comment (nickhsine) [inline src/actions/__test__/header.test.js]: delete this file?

**Comment by YuCJ** [comment]
Will merge this PR and release v5.0.0 after testing beta version on staging of `twreporter-react`

**Comment by YuCJ** [comment]
Release `5.0.0-beta.11`

File: src/actions/__test__/posts.test.js [hunk 9]
```diff
@@ -72,6 +74,9 @@ describe('Testing fetchAFullPost:', () => {
             'mock-slug': mockPost,
           },
         },
+        [fieldNames.origins]: {
+          api: 'http://localhost:8080',
```
Comment (nickhsine) [inline src/actions/__test__/posts.test.js]: replace `http://localhost:8080` to `mockApiHost` ?

File: src/actions/bookmarks.js [hunk 10]
```diff
@@ -177,10 +171,10 @@ export function createSingleBookmark(jwt, uesrID, bookmarkToBeCreated) {
...
 export function getMultipleBookmarks(jwt, userID, offset, limit, sort) {
   return function(dispatch, getState, { httpClientWithToken }) {
-    const url = formAPIURL(
-      `${apiEndpoints.getBookmarks(
-        userID
-      )}?offset=${offset}&limit=${limit}&sort=${sort}`,
+    const url = formURL(
+      `/v1/users/${userID}/bookmarks`,
+      { offset, limit, sort },
+      undefined,
```
Comment (nickhsine) [inline src/actions/bookmarks.js]:
```
const url = formURL(
  apiOrigin,
  `/v1/users/${userID}/bookmarks`,
  { offset, limit, sort }, 
  false, 
)
```
?

File: src/actions/bookmarks.js [hunk 11]
```diff
@@ -234,8 +228,10 @@ export function getMultipleBookmarks(jwt, userID, offset, limit, sort) {
  */
 export function getSingleBookmark(jwt, userID, bookmarkSlug, bookmarkHost) {
   return function(dispatch, getState, { httpClientWithToken }) {
-    let url = formAPIURL(
-      `${apiEndpoints.getSingleBookmark(userID, bookmarkSlug, bookmarkHost)}`,
+    const url = formURL(
+      `/v1/users/${userID}/bookmarks/${bookmarkSlug}`,
+      { host: bookmarkHost },
+      undefined,
```
Comment (nickhsine) [inline src/actions/bookmarks.js]: See above comment.

File: src/actions/bookmarks.js [hunk 12]
```diff
@@ -288,8 +284,10 @@ export function getSingleBookmark(jwt, userID, bookmarkSlug, bookmarkHost) {
  */
 export function deleteSingleBookmark(jwt, userID, bookmarkID) {
   return function(dispatch, getState, { httpClientWithToken }) {
-    const url = formAPIURL(
-      apiEndpoints.deleteSingleBookmark(userID, bookmarkID),
+    const url = formURL(
+      `/v1/users/${userID}/bookmarks/${bookmarkID}`,
+      {},
+      undefined,
```
Comment (nickhsine) [inline src/actions/bookmarks.js]: see above comment.

File: src/store/create-store.js [hunk 13]
```diff
@@ -44,27 +44,29 @@ export default async function createStore(
...
       }
       // Subscribe the redux store changes.
       // Sync the browser storage after redux state change.
+      // Use `debounce` (instead of throttle) to save the last result in duration into storage
       store.subscribe(
-        _.throttle(() => {
+        _.debounce(() => {
           bs.syncReduxState(store.getState())
-        }, 1000)
+        }, 3000)
```
Comment (nickhsine) [inline src/store/create-store.js]: `debounce` is ok for me as well.
However, [the main difference between throttle and debouncing is that throttle guarantees the execution of the function regularly, at least every X milliseconds](https://css-tricks.com/debouncing-throttling-explained-examples/#article-header-id-5).

File: src/utils/browser-storage.js [hunk 14]
```diff
@@ -101,10 +101,12 @@ async function setReduxStateExpires(maxAge = 600) {
  *
  * @returns {Promise<Object>}  A Promise, resolve with a redux state copy in the browser storage
  */
-function getReduxState() {
+function getStateFromStorage() {
   if (!detectEnv.isBrowser()) {
     return Promise.reject(
-      new Error('getReduxState function should be executed on client side')
+      new Error(
+        '`getStateFromStorage` function should be executed on client side'
```
Comment (nickhsine) [inline src/utils/browser-storage.js]: `browser-storage.getStateFromStorage` ?

**Comment by nickhsine** [comment]
LGTM 👍 
Hard work 🔥 🔥 🔥

**Pull request merged**